### PR TITLE
Run multiple benchmarks per node with GROMACS `-multidir`

### LINKED ISSUE #167: Multiple simulations per node
Running multiple simulations per node is beneficial when the node contains GPUs. We want to include this functionality in `version-3` and release it soon. This issue should outline the functionality and some considerations required for the implementation.

# Generate
  - Add new `--multidir` option.
  - Restrict `--min-nodes` and `--max-nodes` to `1`. Warn the user that their input will be ignored.
  - Generate `Treant` in parent folder and separate `Treant` for every multidir child repetition.
    - Parent folder has `multidir=True` category, whereas each child has `multidir_child=True`.
  - Add `_multidir` to folder with engine version, i.e., `2018.6_gpu_multidir`.
# Analyze
  - Ignore parent `Treant` for analysis, look only in child `Treant` for `md.log` file.
  - Calculate average of all child `md.log` files for performance.
  - Print one line for the average in the console.
# Plot
  - Same considerations as for `analyze`.
  - Add a new plot type to showcase the differences between 1 sim./node, 2 sim./node and 4 sim./node?
  - Different plotting types for either 1 sim./node or multidir data?

### PR DESCRIPTION
**Pull request by ksebaz**
Fixes #167

Changes made in this pull request:
- add multidir option to allow for multiple simulations (currenly the gromacs '-multidir' option) is supported

PR Checklist
------------
 - [x] Added changelog fragment in  `./changelog/` ([more information](https://github.com/bio-phys/MDBenchmark/blob/master/DEVELOPER.rst#using-towncrier))?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: mdbenchmark/mdengines/namd.py [hunk 1]
```diff
@@ -28,6 +28,9 @@
 def prepare_benchmark(name, relative_path, *args, **kwargs):
     benchmark = kwargs["benchmark"]
 
+    if not kwargs["multidir"] == 1:
+        raise NotImplementedError
```
Comment (mimischi) [inline mdbenchmark/mdengines/namd.py:32]: This is OK for now, but we should rather use our own `console.error` here. We don't want to throw a Python error at the user and instead just exit with an error message.

File: mdbenchmark/mdengines/namd.py [hunk 2]
```diff
@@ -49,6 +52,8 @@ def prepare_benchmark(name, relative_path, *args, **kwargs):
 
     return name
 
+def prepare_multidir(multidir):
+    raise NotImplementedError
```
Comment (mimischi) [inline mdbenchmark/mdengines/namd.py:56]: Same as above

File: mdbenchmark/mdengines/gromacs.py [hunk 3]
```diff
@@ -19,11 +19,13 @@
 # along with MDBenchmark.  If not, see <http://www.gnu.org/licenses/>.
 import os
 from shutil import copyfile
+import string
 
 from mdbenchmark import console
 
 NAME = "gromacs"
 
+MULTIDIRS=string.ascii_lowercase
```
Comment (mimischi) [inline mdbenchmark/mdengines/gromacs.py:28]: Let's be more specific here and use "LOWERCASE_LETTERS" to make the usage more explicit.

```suggestion
LOWERCASE_LETTERS = string.ascii_lowercase
```

File: mdbenchmark/mdengines/gromacs.py [hunk 4]
```diff
@@ -35,10 +37,23 @@ def prepare_benchmark(name, relative_path, *args, **kwargs):
 
     filepath = os.path.join(relative_path, full_filename)
 
-    copyfile(filepath, benchmark[full_filename].relpath)
+    if kwargs["multidir"] == 1:
+        copyfile(filepath, benchmark[full_filename].relpath)
+    else:
+        for i in range(kwargs["multidir"]):
+            subdir = benchmark[MULTIDIRS[i]+"/"+full_filename].make()
```
Comment (mimischi) [inline mdbenchmark/mdengines/gromacs.py:44]: I assume our linters will complain about the style here. Instead it probably should read:

```suggestion
            subdir = benchmark[LOWERCASE_LETTERS[i] + "/" + full_filename].make()
```

File: mdbenchmark/mdengines/gromacs.py [hunk 5]
```diff
@@ -35,10 +37,23 @@ def prepare_benchmark(name, relative_path, *args, **kwargs):
...
     return name
 
+def prepare_multidir(multidir):
+    if multidir == 1:
+        return ""
+    else:
+        multidir_string = "-multidir"
+        for i in range(multidir):
+            multidir_string += " "+MULTIDIRS[i]
+        return multidir_string
```
Comment (mimischi) [inline mdbenchmark/mdengines/gromacs.py:56]: We can rearrange this code to only return once.

```suggestion
    multidir_string = ""
    
    if multidir != 1:
        multidir_string = "-multidir"
        for i in range(multidir):
            multidir_string += " " + LOWERCASE_LETTERS[i]
            
    return multidir_string
```

But what I do not understand, why are we defining `multidir_string = "-multidir"`, instead of `"_multidir"`?

File: mdbenchmark/mdengines/utils.py [hunk 6]
```diff
@@ -101,20 +101,29 @@ def analyze_benchmark(engine, benchmark):
     threads = np.nan
     hyperthreading = np.nan
     module = None
+    multidir = np.nan
 
     # search all output files
     output_files = glob(
         os.path.join(benchmark.relpath, PARSE_ENGINE[engine.NAME]["analyze"])
     )
     if output_files:
```
Comment (mimischi) [inline mdbenchmark/mdengines/utils.py:110]: I realize this line is not part of the PR, but we can actually remove the `if`-clause. The `for`-loop on the next line will not be executed on an empty list (which is returned when `glob` does not find any matches).

File: mdbenchmark/mdengines/utils.py [hunk 7]
```diff
@@ -101,20 +101,29 @@ def analyze_benchmark(engine, benchmark):
...
-        with open(output_files[0]) as fh:
-            performance = parse_ns_day(engine, fh)
-            fh.seek(0)
-            ncores = parse_ncores(engine, fh)
+        for f in output_files:
+            performance = []
+            ncores = []
```
Comment (mimischi) [inline mdbenchmark/mdengines/utils.py:113]: You need to put the arrays outside the `for`-loop as they will be redefined with every file, no?
```suggestion
        performance = []
        ncores = []
        for f in output_files:
```

File: mdbenchmark/versions.py [hunk 8]
```diff
@@ -192,6 +199,7 @@ class Version3Categories(VersionCategories):
         "hyperthreading": "Hyperthreading?",
         "job_name": "Job name",
         "submitted": "Submitted?",
+        "multidir": "# sims",
```
Comment (mimischi) [inline mdbenchmark/versions.py:202]:
```suggestion
        "multidir": "# Simulations",
```

**Review by mimischi — COMMENTED** [review]
This looks good! I will pull the code and take a look at the functionality now. I made a few comments to improve the readability, but overall this looks good!

File: mdbenchmark/mdengines/utils.py [hunk 9]
```diff
@@ -101,20 +101,29 @@ def analyze_benchmark(engine, benchmark):
...
+            with open(f) as fh:
+                performance.append(parse_ns_day(engine, fh))
+                fh.seek(0)
+                ncores.append(parse_ncores(engine, fh))
+        performance = np.mean(performance)
```
Comment (mimischi) [inline mdbenchmark/mdengines/utils.py:118]: We would want to form the sum not the average of performance values. Running `-multidir` will most of the time, produce shorter individual trajectories. What the user will be interested in is: can I increase my cumulative performance by running multiple simulations per node?

```suggestion
        performance = np.sum(performance)
```

**Review by mimischi — COMMENTED** [review]
Added one more comment!

File: mdbenchmark/mdengines/namd.py (hunk 2 above)
Comment (ksebaz) [reply mdbenchmark/mdengines/namd.py:56]: The error here is actually not correct i think, because it will cause the namd generation to crash by default. Since there is no "multidir" string to be replaced in the namd-template, here an empty string is actually the correct retrun value.

File: mdbenchmark/mdengines/gromacs.py (hunk 5 above)
Comment (ksebaz) [reply mdbenchmark/mdengines/gromacs.py:56]: yes that is much better!

It is "-multidir" because that is being inserted in the template, so it is the gromacs option.

Comment (mimischi) [reply mdbenchmark/mdengines/gromacs.py:56]: Now I understand, OK!

File: mdbenchmark/mdengines/namd.py (hunk 1 above)
Comment (ksebaz) [reply mdbenchmark/mdengines/namd.py:32]:
```suggestion
        raise console.error("The NAMD-engine currently only supports '--multidir 1'")
```

File: mdbenchmark/mdengines/namd.py (hunk 2 above)
Comment (ksebaz) [reply mdbenchmark/mdengines/namd.py:56]: ... or maybe better to return `None`
```suggestion
    return None
```

File: mdbenchmark/mdengines/utils.py (hunk 7 above)
Comment (ksebaz) [reply mdbenchmark/mdengines/utils.py:113]: yes of course! Sorry for that!

**Comment by mimischi** [comment]
@ksebaz Our CI suite is not running with the current branch. I made a small change in the original `version-3` branch. Could you rebase with `git rebase -i version-3` and force-push via `git push origin --force`? This should trigger the CI.

`mdbenchmark generate` seems to work fine, but `mdbenchmark analyze` throws an error. Maybe the CI suite will give us some insights.

**Comment by ksebaz** [comment]
Do you really want me to rebase? Why not just merge in version-3?

**Comment by mimischi** [comment]
It's just a personal preference. I like rebasing because it does not create an additional merge commit. You can also merge the latest changes from `version-3` into your branch if you want!

**Comment by ksebaz** [comment]
> It's just a personal preference. I like rebasing because it does not create an additional merge commit. You can also merge the latest changes from version-3 into your branch if you want!
Ok I did that. The pipeline seems to start

I haven't looked at the testcases so far since I first wanted your general approval, but I am already doing that now...

**Comment by ksebaz** [comment]
At the moment the MacOS stage is still failing, but that is unrelated to the changes here, but due to some issues with python versions.

File: mdbenchmark/mdengines/namd.py [hunk 10]
```diff
@@ -28,6 +28,9 @@
 def prepare_benchmark(name, relative_path, *args, **kwargs):
     benchmark = kwargs["benchmark"]
 
+    if not kwargs["multidir"] == 1:
+        raise console.error("The NAMD-engine currently only supports '--multidir 1'")
```
Comment (mimischi) [inline mdbenchmark/mdengines/namd.py:32]: We do not need to raise `console.error`. Calling it will ultimately call `sys.exit()` and exit gracefully.

```suggestion
        console.error("The NAMD-engine currently only supports '--multidir 1'")
```

**Review by mimischi — APPROVED** [review]
One more minor thing. Otherwise, this looks good to me!

I will have a look at the macOS error. I've had the impression that I have solved this already.

File: mdbenchmark/mdengines/namd.py (hunk 10 above)
Comment (ksebaz) [reply mdbenchmark/mdengines/namd.py:32]: uups! sorry I knew that actually.

**Comment by mimischi** [comment]
MDBenchmark remains to crash for me, when I try to run `mdbenchmark analyze`. I've installed the latest version from this PR. It looks like our unit tests are not covering a specific case? I guess the problem was there before your PR!

```
mdbenchmark generate --name md --module gromacs/2018.8 --skip-validation --host cobra --multidir 4
mdbenchmark analyze

Setting up...
Traceback (most recent call last):
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/pandas/core/internals/construction.py", line 500, in _list_to_arrays
    content, columns, dtype=dtype, coerce_float=coerce_float
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/pandas/core/internals/construction.py", line 583, in _convert_object_array
    "{con} columns".format(col=len(columns), con=len(content))
AssertionError: 11 columns passed, passed data had 12 columns

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/bin/mdbenchmark", line 11, in <module>
    load_entry_point('mdbenchmark', 'console_scripts', 'mdbenchmark')()
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/migecht/Code/MDBenchmark/mdbenchmark/cli/commands.py", line 69, in analyze
    do_analyze(directory=directory, save_csv=save_csv)
  File "/Users/migecht/Code/MDBenchmark/mdbenchmark/cli/analyze.py", line 35, in do_analyze
    bundle, columns=version.analyze_categories, sort_values_by=version.analyze_sort,
  File "/Users/migecht/Code/MDBenchmark/mdbenchmark/utils.py", line 223, in parse_bundle
    df = pd.DataFrame(data, columns=columns)
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/pandas/core/frame.py", line 450, in __init__
    arrays, columns = to_arrays(data, columns, dtype=dtype)
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/pandas/core/internals/construction.py", line 464, in to_arrays
    return _list_to_arrays(data, columns, coerce_float=coerce_float, dtype=dtype)
  File "/Users/migecht/Library/Caches/pypoetry/virtualenvs/mdbenchmark-IJNzBYdu-py3.7/lib/python3.7/site-packages/pandas/core/internals/construction.py", line 503, in _list_to_arrays
    raise ValueError(e) from e
ValueError: 11 columns passed, passed data had 12 columns
```

**Comment by ksebaz** [comment]
That is strange. For me that exact same command works without errors and as expected.

**Comment by mimischi** [comment]
My bad! I should stop testing in the root of the codebase. `mdbenchmark analyze` was picking up folders from the `mdbenchmark/tests/` directory. This looks like it works!

Is there anything you want to add @ksebaz?

**Comment by ksebaz** [comment]
the error message for the pandas-dataframe creation is irritating. The data should actually have 12 columns and that is also reporten. But also the columns which come from analyze_categories should have 12 columns in version 3.  And in the PR they also have that. So why is your error saying 11 columns. Is there something wrong with the installation?

**Comment by ksebaz** [comment]
Ah ok, that makes sense

**Comment by ksebaz** [comment]
> Is there anything you want to add @ksebaz?

I am currently running some consistency checks comparing the version-3 branch against multidir branch with "--multidir 1". I let you know when that is finished ...

**Comment by ksebaz** [comment]
> I am currently running some consistency checks comparing the version-3 branch against multidir branch with "--multidir 1". I let you know when that is finished ...

That looks fine. Non inconsistencies found.

**Comment by mimischi** [comment]
Thank you a lot @ksebaz! I think we are good to go. We can ignore the Travis CI error for now, as it is not caused by this PR.

Is there anything else you want to add? If not, I would go ahead and merge the changes!

**Comment by ksebaz** [comment]
> Is there anything else you want to add? If not, I would go ahead and merge the changes!

I have no more changes at the moment. So from my side you can merge.
